fix the watch cache

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

p0lyn0mial committed:
UPSTREAM: <carry>: the watch cache needs to recalculate the key for an object

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:
```diff
@@ -277,6 +279,26 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 		return fmt.Errorf("couldn't compute key: %v", err)
 	}
+
+	{ // hack
```
I'll have to rework it, just to get it going.

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1125,7 +1127,8 @@ func (lw *cacherListerWatcher) List(options metav1.ListOptions) (runtime.Object,
 	}
 
-	if err := lw.storage.List(context.TODO(), lw.resourcePrefix, storage.ListOptions{ResourceVersionMatch: options.ResourceVersionMatch, Predicate: pred}, list); err != nil {
+	ctx := genericapirequest.WithCluster(context.TODO(), genericapirequest.Cluster{Name: logicalcluster.Wildcard, Wildcard: true})
```
it allows us to bring all available data

p0lyn0mial commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:
```diff
@@ -277,6 +279,26 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
+			key = fmt.Sprintf("%v/%v/%v", strings.Join(sk[:len(sk)-2], "/"), clusterName, strings.Join(sk[len(sk)-2:], "/"))
+		} else if len(clusterName) > 0 {
+			sk := strings.Split(key, "/")
+			key = fmt.Sprintf("%v/%v/%v", strings.Join(sk[:len(sk)-1], "/"), clusterName, strings.Join(sk[len(sk)-1:], "/"))
```
recomputing the key under we store objects will be necessary

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:
```diff
@@ -277,6 +279,26 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
+		default:
+			klog.Infof("could not get ClusterName on object: %T", event.Object)
+		}
+		if len(ns) > 0 && len(clusterName) > 0 {
```
Can we get a test or at least a comment with expected inputs and outputs? This is a bit hard to read with all the splits and joins

stevekuznetsov replied:
Also was the reason we needed to do this ... that watch cache is computing keys in some other way from the keyfunc in listers?

p0lyn0mial replied:
done. The reflector used by the watch cache uses `logicalcluster. Wildcard` to discover all resources for the given GVR. During that process, the clusterName will be assigned by the storage layer. The watch cache needs to rewrite the URLs for each resource to contain the clusterName as this is how the clients will ask for resources.

p0lyn0mial committed:
UPSTREAM: <carry>: allow for passing a context to the key fn that is used by a storage decorator

p0lyn0mial committed:
UPSTREAM: <carry>: introduce KcpStorageMetadata for holdng KCP specific metadata

it is used by the watch cache's reflector to instruct the storage layer how to assign/extract the cluster name

p0lyn0mial commented:
this PR is ready for review PTAL. Thanks!

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/context_cluster.go:
```diff
@@ -94,3 +101,10 @@ func ClusterNameFrom(ctx context.Context) (logicalcluster.Name, error) {
 	return cluster.Name, nil
 }
+
+func IsFromStorageDecorator(ctx context.Context) bool {
```
Go doc, waht does this mean?

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/request/context_cluster.go:
```diff
@@ -29,6 +29,9 @@ type clusterKey int
 	clusterContextKey clusterKey = iota
+
+	// cacheWatermarkContextKey is the context key to indicate the key function is being used by a storage decorator
+	storageDecoratorsWatermarkContextKey clusterKey = iota
```
what does it mean? By a decorator? Can you expand a bit in the comment?

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -240,7 +240,9 @@ func NoNamespaceKeyRootFunc(ctx context.Context, prefix string) string {
-		klog.Errorf("invalid context cluster value: %v", err)
+		if !genericapirequest.IsFromStorageDecorator(ctx) {
+			klog.Errorf("invalid context cluster value: %v", err)
+		}
```
worth to mark as KCP change?

p0lyn0mial replied:
what do you mean by that?

sttts commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1421,11 +1423,15 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
 		if isNamespaced {
-			return e.KeyFunc(genericapirequest.WithNamespace(genericapirequest.NewContext(), accessor.GetNamespace()), accessor.GetName())
+			return e.KeyFunc(genericapirequest.WithNamespace(ctx, accessor.GetNamespace()), accessor.GetName())
 		}
```
worth to mark as KCP change?

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -96,6 +99,9 @@ type Config struct {
 	Clock clock.Clock
+
+	// KcpIndicators hold some KCP indicators that are used by the reflector to instruct the storage layer how to extract the cluster name
+	KcpIndicators context.Context
```
what is an indicator?

p0lyn0mial replied:
I use this term for metadata that are needed by the storage layer. I can rename to `KcpStorageLayerMetadata`

sttts replied:
KcpExtraStorageMetadata?

Can you give examples here in the go doc ?

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go:
```diff
@@ -0,0 +1,63 @@
+package cacher
```
patch_watch_cache.go is our pattern, isn't it?

p0lyn0mial replied:
I think it is correct, we have `store_patch.go`, here:
https://github.com/kcp-dev/kubernetes/blob/a8b2000cd5878fb3d0b8dbc74bec068c0a034a8e/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:
```diff
@@ -99,6 +100,11 @@ type ConfigForResource struct {
+	// The following fields hold config required/specific to KCP
+	//
+	// KcpIndicators hold some KCP indicators that are used by the reflector to instruct the storage layer how to extract the cluster name
+	KcpIndicators context.Context
```
what is an indicator on a context? Having a ctx here looks strange. Convince me it isn't :-)

p0lyn0mial replied:
sure, I can make it a struct, will be more explicit.

p0lyn0mial replied:
essentially a struct that will hold some metadata for the storage layer - passed by the reflector.

sttts replied:
but why in a context? Why not typed?

p0lyn0mial replied:
no specific reason, let's make it more concrete, I will change it to a struct.

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go:
```diff
@@ -0,0 +1,63 @@
+
+// recalculateWatchCacheKey adjusts the key under which the given object will be stored in the watch cache.
+//
+// resources in the db are stored without the clusterName, since the reflector used by the cache uses the logicalcluster.Wildcard
```
This is a touch hard to follow - all objects get their clusterName assigned by the storage layer, regardless of wildcard or not. Is the cache pre-calculating the key in a non-cluster-aware manner, and that's why we are updating a previous key here, as opposed to just generating it correctly in the first place?

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go:
```diff
@@ -0,0 +1,63 @@
+	case metav1.ObjectMetaAccessor:
+		clusterName = t.GetObjectMeta().GetClusterName()
+		ns = t.GetObjectMeta().GetNamespace()
+	case clusterNameAndNsGetter:
```
What fulfills this interface and not ObjectMetaAccessor?

p0lyn0mial replied:
by `unstructured.Unstructured`

stevekuznetsov replied:
Maybe worth a comment, I am surprised that it's true

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go:
```diff
@@ -0,0 +1,63 @@
+	//  /group/resource/namespace/instance => /group/resource/clusterName/namespace/instance
+	//
+	//  /group/resource/$identity/namespace/instance => /group/resource/$identity/clusterName/namespace/instance
+	if len(ns) > 0 {
```
To what extent do we need to control edge case behavior here? We haven't checked that the `ns` exists in the key yet, nor that it is at the 2nd position by path separators

stevekuznetsov commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1421,11 +1423,15 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
 			return "", err
 		}
 
+		// keyFunc is used by the storage decorators and there is no way to pass a context
```
What does it mean "no way to pass a context" - aren't you passing contexts in the next lines? :)

p0lyn0mial replied:
actually, I have just realized we should change all the layers to accept the context. That would simplify this PR and allow to reuse of some code. One sec :)

p0lyn0mial replied:
@stevekuznetsov pushed, ptal

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1101,17 +1115,19 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
+		storage:                storage,
+		resourcePrefix:         resourcePrefix,
+		newListFunc:            newListFunc,
+		kcpClusterAwareContext: createKCPClusterAwareContext(kcpExtraStorageMetadata),
```
storing a context still feels strange. Why not store `kcpExtraStorageMetadata` and call `createKCPClusterAwareContext` further down?

p0lyn0mial replied:
storing a ctx seems easier here because I need to create the context only once and reuse it multiple times

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -1140,7 +1156,15 @@ func (lw *cacherListerWatcher) Watch(options metav1.ListOptions) (watch.Interfac
+	return lw.storage.WatchList(lw.createKCPClusterAwareContext(), lw.resourcePrefix, opts)
+}
+
+func (lw *cacherListerWatcher) createKCPClusterAwareContext() context.Context {
```
nit: `func createKCPClusterAwareContext(meta *storagebackend.KcpStorageMetadata)`.

rule-of-thumb: functional (= non-side-effect) funcs should not be methods.

sttts commented:
/lgtm
/approve
